src/endorse/common/memoize2.py

```diff
@@ -0,0 +1,238 @@
+    The function code is inspected and referenced other functions are hashed if
+    a) they are from the same module
+    b) they are decorated by memoize itself
+    :param cache: A caching class providing 'insert' and 'value' methods.
+    Simple implementation is provided by ResultCacheFile. Storing pickled results in separate files.
+    Not suitable for many small results.
+    :return: decorated function
+    """
+    def decorator(func):
+        cloudpickle.register_pickle_by_value(sys.modules[func.__module__])
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            m = hashlib.sha256()
+            pickled_fn_call = cloudpickle.dumps((func, args, kwargs))
+            m.update(pickle_remove_duplicit(pickled_fn_call))
+            hash_fn_call = m.digest()
+
+            fun_name = func.__name__
+            value = cache.value(hash_fn_call, fun_name)
+            if value is ResultCacheFile.NoValue:
+                result = func(*args, **kwargs)
+                data = {
+                    "func": func,
+                    "args": args,
+                    "kwargs": kwargs,
+                    "result": result
+                }
+                cache.insert(hash_fn_call, data, fun_name)
+            else:
+                result = value["result"]
+                if _config["debug"]:
+                    new_result = func(*args, **kwargs)
+                    if new_result != result:
+                        # Detection of faulty hashing.
+                        # TODO: better reporting
+                        print("Function {}, new result is different from cache result.".format(fun_name))
+
+            return result
+
+        wrapper.__name__ += "_memoize"
+
+        return wrapper
+    return decorator
+
+
+def pickle_remove_duplicit(pkl):
+    """
+    Optimize a pickle microcode by removing unused PUT opcodes and duplicated unicode strings.
+    This is critical to avoid spurious hash differences.
+    """
+    put = 'PUT'
+    get = 'GET'
+    oldids = set()  # set of all PUT ids
+    newids = {}     # set of ids used by a GET opcode; later used to map used ids.
+    opcodes = []    # (op, idx) or (pos, end_pos)
+    proto = 0
+    protoheader = b''
+    strings = {}
+    memo_map = {}
+    last_opcode_name = ""
+    last_arg = None
+
+    # Generate all opcodes and store positions to calculate end_pos
+    ops = list(pickletools.genops(pkl))
+    for i, (opcode, arg, pos) in enumerate(ops):
+        # Determine end_pos by looking at the position of the next opcode
+        end_pos = ops[i + 1][2] if i + 1 < len(ops) else len(pkl)
+
+        if 'PUT' in opcode.name:
+            assert opcode.name in ('PUT', 'BINPUT'), f"{opcode.name}"
+            oldids.add(arg)
+            opcodes.append((put, arg))
+        elif opcode.name == 'MEMOIZE':
+            idx = len(oldids)
+
+            # Inserted into optimize
+            if 'BINUNICODE' in last_opcode_name:
```

Comment:
Je jisté, že MEMOIZE bude hned za BINUNICODE se stringem? 
Nebylo by lepší detekovat 'UNICODE' in last_opcode_name or 'STRING' in last_opcode_name ...
kdyby někdo změnil implementaci uložení problematických cest. Možná by šlo prostě porovnat typ argumentu: isinstance(last_arg, str), nebo pro jistotu deduplikovat všechno.